nbconvert/preprocessors/execute.py

```diff
@@ -81,6 +81,40 @@ def from_cell_and_msg(cls, cell, msg):
 {ename}: {evalue}
 """
 
+class CellExecutionTimeoutError(TimeoutError):
+    """
+    Custom exception to propagate exceptions that are raised during
+    notebook execution to the caller for timeout. This is mostly
+    useful when using nbconvert as a library, since it allows to
+    deal with failures gracefully.
+    """
+    def __init__(self, traceback):
+        super(CellExecutionTimeoutError, self).__init__(traceback)
+        self.traceback = traceback
+
+    def __str__(self):
+        s = self.__unicode__()
+        if not isinstance(s, str):
+            s = s.encode('utf8', 'replace')
+        return s
+
+    def __unicode__(self):
+        return self.traceback
+
+    @classmethod
+    def from_cell_and_msg(cls, cell):
+        """Instantiate from a code cell object
+        (message is timeout when executing cell)
+        """
+        return cls(exec_timeout_err_msg.format(cell=cell))
+
+exec_timeout_err_msg = u"""\
+Timed out while executing the following cell:
+------------------
+{cell.source}
```

Comment:
The source could be quite long - I believe it was suggested to just take the `head` and `tail` if the source is >10 lines long.